### PR TITLE
Add the palette_generator package to the packages repo

### PR DESCRIPTION
**Pull request by gspencergoog**
This adds a palette generator that will take an image and generate a set of distinct, dominant colors from it.  This is more or less a port of the [Android Palette](https://android.googlesource.com/platform/frameworks/support/+/android-8.1.0_r33/v7/palette/src/main/java/android/support/v7/graphics/Palette.java) class, except that it doesn't automatically resize the image if it is too large.

screenshot_1529385906

Where the first (top left below the image) color is the more prominent distinct color, and the last is the least. It can select various target colors from the available colors based on targets that it is given (and has several default targets it looks for).

The screenshot above is from the sample app that shows how to use it, and lets the user select a region of the image to take the palette from.

[Also, this PR includes the CI PR #1, but I'll merge and resolve that before committing this one.]

### CODE REVIEW

**Comment by gspencergoog** [comment]
This is ready for review.  The Fuschia folks would like it comitted by Wednesday for a demo in Paris on Thursday, June 20th, if at all possible.

**Comment by lukepighetti** [comment]
I would have immediate use for this. Glad to see that someone is working on it. :)

File: packages/palette_generator/example/image_colors/lib/main.dart [hunk 1]
```diff
@@ -0,0 +1,311 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+//import 'dart:async';
+//import 'dart:math' as math;
+
+//import 'package:palette_generator/palette_generator.dart';
+
+//import 'package:flutter/gestures.dart';
```
Comment (Hixie) [inline packages/palette_generator/example/image_colors/lib/main.dart]: commented out code?

File: packages/palette_generator/example/image_colors/lib/main.dart [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+            label: 'Light Vibrant', color: generator.lightVibrantColor?.color),
+        new PaletteSwatch(
+            label: 'Vibrant', color: generator.vibrantColor?.color),
+        new PaletteSwatch(
+            label: 'Dark Vibrant', color: generator.darkVibrantColor?.color),
+        new PaletteSwatch(
+            label: 'Light Muted', color: generator.lightMutedColor?.color),
+        new PaletteSwatch(label: 'Muted', color: generator.mutedColor?.color),
+        new PaletteSwatch(
+            label: 'Dark Muted', color: generator.darkMutedColor?.color),
```
Comment (Hixie) [inline packages/palette_generator/example/image_colors/lib/main.dart]: having some of these wrapped and some not makes this hard to read

File: packages/palette_generator/lib/palette_generator.dart [hunk 3]
```diff
@@ -0,0 +1,1180 @@
...
+  PaletteColor get mutedColor => selectedSwatches[PaletteTarget.muted];
+
+  /// Returns a muted and light color from the palette. Might be null if an
+  /// appropriate target color could not be found.
+  PaletteColor get lightMutedColor =>
+      selectedSwatches[PaletteTarget.lightMuted];
+
+  /// Returns a muted and dark color from the palette. Might be null if an
+  /// appropriate target color could not be found.
+  PaletteColor get darkMutedColor => selectedSwatches[PaletteTarget.darkMuted];
```
Comment (Hixie) [inline packages/palette_generator/lib/palette_generator.dart]: ditto with the inconsistent wrapping

File: packages/palette_generator/lib/palette_generator.dart [hunk 4]
```diff
@@ -0,0 +1,1180 @@
...
+  bool _shouldBeScoredForTarget(
+      PaletteColor paletteColor, PaletteTarget target, Set<Color> usedColors) {
+    // Check whether the HSL lightness is within the correct range, and that
+    // this color hasn't been used yet.
+    final HSLColor hslColor = new HSLColor.fromColor(paletteColor.color);
+    return hslColor.saturation >= target.minimumSaturation &&
+        hslColor.saturation <= target.maximumSaturation &&
+        hslColor.lightness >= target.minimumLightness &&
+        hslColor.lightness <= target.maximumLightness &&
+        !usedColors.contains(paletteColor.color);
```
Comment (Hixie) [inline packages/palette_generator/lib/palette_generator.dart]: inconsistent indentation

File: packages/palette_generator/lib/palette_generator.dart [hunk 5]
```diff
@@ -0,0 +1,1180 @@
...
+  bool operator ==(dynamic other) {
+    return minimumSaturation == other.minimumSaturation &&
+        targetSaturation == other.targetSaturation &&
+        maximumSaturation == other.maximumSaturation &&
+        minimumLightness == other.minimumLightness &&
+        targetLightness == other.targetLightness &&
+        maximumLightness == other.maximumLightness &&
+        saturationWeight == other.saturationWeight &&
+        lightnessWeight == other.lightnessWeight &&
+        populationWeight == other.populationWeight;
```
Comment (Hixie) [inline packages/palette_generator/lib/palette_generator.dart]: ditto

**Comment by gspencergoog** [comment]
I left the indentation issues because this repo has the dartfmt presubmit check turned on like flutter/plugins does (and so I can't change those without putting in a formatting exception comment).

**Comment by gspencergoog** [comment]
Hi @quentinleguennec Yes, unfortunately, it requires some fixes in the engine that are not yet available in the dev/beta releases, so it only works with the master channel right now. We're in the process of rolling to dev, so soon it will "just work", but until then you'll either have to use the master channel, or fork the code.

If I remember correctly, the issue was only with the example app and testing code, not with the palette generator itself, but those same engine bugs might make it hard for you to use the palette generator (depending on the source of your images). They had to do with getting raw pixels from a `ui.Image`.

**Comment by quentinleguennec** [comment]
Ok, thanks for the info. Do you have an idea of when it should make it to beta?

**Comment by gspencergoog** [comment]
It will be in the next dev roll, and then the next beta roll, whenever that completes (It depends on a quality metric that has to be met, so I can't give you a specific date). I would estimate sometime in the next month.

**Comment by quentinleguennec** [comment]
Thanks :+1:

**Comment by gromdroid** [comment]
I'm trying to use the palettegenerator inside a Isolate but it never returns anything. I'm trying to calculate the dominant color for a bunch of images and it's causing the UI to lag. Because of this I tried to offload the task to an Isolate but for some reason the palettegenerator future never gives a result in an Isolate. Any idea how this can be done?

**Pull request merged**